### PR TITLE
soft-code Postgres in json-api

### PR DESCRIPTION
**Pull request by S11001001**
We turn the hard-coded implicits and queries into references to values that can be configured while we are initializing json-api.

This _does not, and will not in the foreseeable future_ imply that you can "plug in" a different database, _even if you implement what look like abstract interfaces here_. It is meant to allow _json-api itself_ to be more easily extended with support for different databases.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by cocreature — APPROVED** [review]
looks very sensible, thanks!

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala [hunk 1]
```diff
@@ -51,11 +51,22 @@ object WebSocketService {
...
       unresolved: Set[domain.TemplateId.OptionalPkg],
       fn: domain.ActiveContract[LfV] => Option[Positive],
-      dbQuery: OneAnd[Set, domain.Party] => LogHandler => ConnectionIO[
+      dbQuery: (OneAnd[Set, domain.Party], dbbackend.ContractDao) => ConnectionIO[
         _ <: Vector[(domain.ActiveContract[JsValue], Positive)]
       ],
   )
 
+  /** If an element satisfies `prefix`, consume it and emit the result alongside
+    * the next element (which is not similarly tested); otherwise, emit it.
```
[the remark this answers is not in the document]
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala:60]: Tested with `prefix`.

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/SupportedJdbcDriver.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.http.dbbackend
+
+import doobie.{Get, Put}
+
+/** A nullary typeclass of incompatible JDBC operations and settings, selected by
+  * the options passed to json-api at startup.
+  *
+  * We currently use the fact that it is nonsensical to run json-api against more
```
[the remark this answers is not in the document]
Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/SupportedJdbcDriver.scala:11]: Or in a running instance, as you say.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 3]
```diff
@@ -20,40 +20,15 @@ import spray.json._
 import cats.instances.list._
 import cats.Applicative
 import cats.syntax.applicative._
-import cats.syntax.foldable._
 import cats.syntax.apply._
 import cats.syntax.functor._
 
-object Queries {
+sealed abstract class Queries {
```
[the remark this answers is not in the document]
Comment (S11001001) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:26]: Indeed.

File: ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala [hunk 4]
```diff
@@ -62,7 +37,7 @@ object Queries {
   private[this] val createContractsTable: Fragment = sql"""
       CREATE TABLE
         contract
-        (contract_id TEXT PRIMARY KEY NOT NULL
+        (contract_id TEXT NOT NULL PRIMARY KEY
```
[the remark this answers is not in the document]
Comment (S11001001) [reply ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala:40]: This idea comes from your replacement for `BIGSERIAL`.

**Pull request merged**